### ISSUE TITLE
Provide an arm or universal build for macOS.

### ISSUE THREAD

**Issue by foldfree** [OpenTabletDriver/OpenTabletDriver#2955]
### Description

It would be great to have a native arm aarch64 build for apple silicon devices running on macOS.

### Acknowledgements

- [X] I have searched the existing issues and this new issue is not a duplicate of any.
- [X] I have written a concise and meaningful title.

**Comment by jamesbt365** [comment]
The x86 builds work perfectly fine with the compatibility layer in MacOS, so this isnt a high priority at the moment.

From previous testing and if I remember correctly these builds needs to be signed on MacOS which would mess with our CI. Somebody correct me on that if I'm wrong.

**Comment by foldfree** [comment]
Not sure if this is what you mean but it is still possible to run unsigned builds on arm macOS by using `xattr -dr com.apple.quarantine /path/to/program.app`

**Comment by jamesbt365** [comment]
And that is extremely inconvenient for users, far less than just packaging x86 builds and only having one binary for all of MacOS.

**Comment by scj643** [comment]
Building from source using the ARM .net works just fine. Also if needed I could codesign it.

**Comment by jamesbt365** [comment]
ARM builds are impossible until 0.7.0.0 due to a problem with our UX, afaik `0.6.x` (the branch all releases are currently built upon) doesn't work.

**Comment by scj643** [comment]
```
Unhandled exception. System.EntryPointNotFoundException: Unable to find an entry point named 'objc_msgSend_stret' in shared library '/usr/lib/libobjc.dylib'.
   at MonoMac.ObjCRuntime.Messaging.CGRect_objc_msgSend_stret(CGRect& retval, IntPtr receiver, IntPtr selector)
   at MonoMac.AppKit.NSScreen.get_Frame()
   at Eto.Mac.Forms.MouseHandler.get_Position()
   at Eto.Forms.Mouse.get_Position()
   at OpenTabletDriver.UX.MainForm.InitializeForm() in /Users/scj643/devel/OpenTabletDriver/OpenTabletDriver.UX/MainForm.cs:line 87
   at OpenTabletDriver.UX.MainForm..ctor() in /Users/scj643/devel/OpenTabletDriver/OpenTabletDriver.UX/MainForm.cs:line 32
   at OpenTabletDriver.UX.App.RunInternal(String platform, CommandLineOptions options) in /Users/scj643/devel/OpenTabletDriver/OpenTabletDriver.UX/App.cs:line 55
   at OpenTabletDriver.UX.App.Run(String platform, String[] args) in /Users/scj643/devel/OpenTabletDriver/OpenTabletDriver.UX/App.cs:line 41
   at OpenTabletDriver.UX.MacOS.Program.Main(String[] args) in /Users/scj643/devel/OpenTabletDriver/OpenTabletDriver.UX.MacOS/Program.cs:line 13
```

Yep the UI just dies :(

Still would be possible to have the daemon be a universal binary though

**Comment by joaofrgomes** [comment]
Hi, I'm the user who just mentioned this issue. Any updates on this front?

I don't have any numbers from proper marketing research on me, but considering how big the performance gap between Intel and Apple Silicon machines was from the get-go and how even bigger it has become since the first M1 machines dropped, pure anecdotal evidence from seeing all my designer buddies canning the latter and adopting the former, and [the announced demise of OCLP](https://github.com/dortania/OpenCore-Legacy-Patcher/issues/1167), which will further force the big old hardware stalwarts (myself included, I'll have to upgrade my daily drivers, demote my oldest Apple Silicon machines to emergency backups, and move all my old Macs, which were indeed my backups, to x86-64 Linux distros, to keep them relevant and useful or something) to start considering upgrades too… there's a perfect storm forming.

Older, unmaintained x86-64 macOS builds and [the imminent demise of Rosetta 2](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment/) will mean that in one, maybe two years at best, x86-64 macOS-only support in OTD will become untenable and make it too much of a niche solution for Mac users, instead of making it **the universal, go-to one**, as Wacom smells the blood on the water from their own forced obsolescence and reaps the benefits from users who'll prioritize having an updated, secure OS on their daily driver over saving on buying a new tablet and inevitably cave in.

Please bear in mind, and I cannot stress this enough, that I'm not hostile or indifferent to your efforts, or to the mindset behind them; I love FOSS solutions and have been using them as much as I possibly can, also contributing with my own beta testing user feedback (Zotero and ODF's LibreOffice immediately come to mind), just asking the hard questions and stating the hard truths here. In that vein, I also would very much like to help you, but preferably on a solution that is guaranteed to work long-term, with the least hiccups possible, and benefit, well, me as well, that being **native Apple Silicon builds of OTD**. I know it's terrible to have to invest on new hardware, address the whole “UX breaking” conundrum (excuse me my ignorance; I'm not too comfortable with having to build stuff from source, so I haven't gone that route just yet), the works, but such is the way of the world…

FWIW, I own a Wacom CTH-670, and will test it today with OTD on the Tahoe RC build I've been running recently (@thenickdude's fixed drivers seem to work fine, and will likely be my preferred solution until they inevitably break, to be honest, but I can see the writing on the wall and welcome my new FOSS overlords).

**Comment by arjun1194** [comment]
hey i have figured out the following
1. Mac os build from source works but the UI does not work for some reason.
2. With that being said you can technically still use `OpenTabletDriver.Daemon` and it works fine for apple silicon builds
3. While the team works on this, i have hacked about a web console that directly controls the `OpenTabletDriver.Daemon` using a html and node js orchestrator script that you can use to start and configure the driver. 

Anyone interested to run apple silicon can use this method as a workaround, i am also available to work on this if the team is not directly working on it. Please let me know if any help is needed

Edit - you can checkout the repo for web UI for `OpenTabletDriver.Daemon` here - https://github.com/arjun1194/open-tablet-driver-web-console

**Comment by floorish** [comment]
I've tried to run the macOS arm build, but couldn't get it to work. Perhaps I'm doing something wrong.

1. Build on Linux (vm), 0.6.x branch (322f261a8)
2. Using `dotnet-sdk-8.0.416-linux-arm64` binary
3. Run`PATH="/path/to/dotnet-sdk-8.0.416-linux-arm64:$PATH" ./build.sh --runtime osx-arm64`
4. The build succeeds, but executing the resulting `OpenTabletDriver.app` or `OpenTabletDriver.app/Contents/MacOS/OpenTabletDriver.Daemon` does not work:

`zsh: killed     ./OpenTabletDriver.app/Contents/MacOS/OpenTabletDriver.Daemon`

The same steps for `--runtime osx-x64` seems to give a working build that runs correctly with Rosetta (both .app and .Daemon). Unfortunately I'm not knowledgable enough to debug this, but just wondering if the build steps should be correct.

EDIT: Of course, right after I posted I read that the .app needs to be code signed on arm (see https://github.com/OpenTabletDriver/OpenTabletDriver/pull/4075). After performing:
`codesign -s - -f --deep ./OpenTabletDriver.app`
Running the `OpenTabletDriver.Daemon` works correctly — however the .app crashes.

**Comment by AkiSakurai** [comment]
> I've tried to run the macOS arm build, but couldn't get it to work. Perhaps I'm doing something wrong.
> 
> 1. Build on Linux (vm), 0.6.x branch ([322f261](https://github.com/OpenTabletDriver/OpenTabletDriver/commit/322f261a81c8f4417653572b4768cb3c131f2368))
> 2. Using `dotnet-sdk-8.0.416-linux-arm64` binary
> 3. Run`PATH="/path/to/dotnet-sdk-8.0.416-linux-arm64:$PATH" ./build.sh --runtime osx-arm64`
> 4. The build succeeds, but executing the resulting `OpenTabletDriver.app` or `OpenTabletDriver.app/Contents/MacOS/OpenTabletDriver.Daemon` does not work:
> 
> `zsh: killed ./OpenTabletDriver.app/Contents/MacOS/OpenTabletDriver.Daemon`
> 
> The same steps for `--runtime osx-x64` seems to give a working build that runs correctly with Rosetta (both .app and .Daemon). Unfortunately I'm not knowledgable enough to debug this, but just wondering if the build steps should be correct.
> 
> EDIT: Of course, right after I posted I read that the .app needs to be code signed on arm (see [#4075](https://github.com/OpenTabletDriver/OpenTabletDriver/pull/4075)). After performing: `codesign -s - -f --deep ./OpenTabletDriver.app` Running the `OpenTabletDriver.Daemon` works correctly — however the .app crashes.

The current version of Eto used in OpenTabletDriver will crash; you have to upgrade Eto.
The latest version of Eto has a bug that causes incorrect autosizing. This bug has already been fixed in the master branch, so we will have to wait for the next Eto release.

**Comment by joaofrgomes** [comment]
Update: Apple is now getting even more aggressive in finishing their transition, by allowing (and strongly encouraging) developers to drop Universal app submissions on their Mac App Store and going for Apple Silicon-only binaries instead, this despite the fact that macOS Tahoe is compatible with Intel machines and will get security updates for two more years, and reminding everyone that “If you haven’t updated your app yet: **Begin the transition immediately** [emphasis mine]. Build and distribute your app as a universal binary, which runs natively on both Apple silicon and Intel-based Mac computers.” on a [new support document](https://developer.apple.com/news/?id=w5ngl9k2).

We still don't know what vestiges of Rosetta 2 will remain or what kind of binaries they'll support, but the same document also states that macOS 27 is the “Final release to support Rosetta — Intel-only apps will no longer run on Mac computers with Apple silicon after this update.”

It doesn't get more obvious and urgent than this and, if anything, this should change the priority on this issue from “low” to something else a bit higher on the latter and prompt at least one of the devs to snatch an old M1 machine (yes, _old_; the first models are three months away from turning six years old at this point, and about to become vintage even by AppleCare standards) from a thrift store and start testing OTD on Apple Silicon machines (including Linux-on-AARCH64 and Windows-on-AARCH64 builds, because a successor or a revival of OCLP/MacPostFactor to run newer builds of macOS on officially unsupported Apple Silicon machines isn't a given and Asahi Linux and Windows on top of some officially revived or community-based Boot Camp-like solution may become the only long-term options for old Macs at some point in the not-so-distant future) in earnest.

I am more than happy to help as an end user with access to an M1 Max Mac Studio and an M2 MacBook Air, but at some point, if you want to create and maintain software that people can actually use indefinitely, you **have** to stay current, because peripherals last and stay useful way longer than desktops and laptops and even people who wish to save the former from the e-waste bin may be forced, for competitive or even security reasons, to upgrade their daily driver machines. There's also the elephant in the room that is Tahoe-on-OCLP; as things stand, it may arrive so late that it won't end up making much of a difference to vintage Mac users now on Sequoia, many of which will prefer to jump ship to maybe an even aging Apple Silicon Mac running Golden Gate or later and reap the performance, security _and_ UX benefits from it (iykyk, but Tahoe has been wildly panned by power and regular users alike as being ugly, badly designed and resource-hungry, and I concur and say this even as I run it on quite zippy Apple Silicon machines and see Sequoia already running as molasses on most of my OCLP machines, so I can't even fathom how horribly it will run on OCLP on anything that isn't an higher end and highly upgraded iMac, iMac Pro or Mac Pro)

TL;DR: old Intel Macs running macOS, no matter how many millions of them are there still alive and kicking today, are getting increasingly irrelevant and unviable as daily drivers (yes, there's a cut-off point and it's fast approaching – two years, or maybe just one if the OCLP team falters! –, and it will be a bummer for a lot of people and organisations that rely on them), and the only way to keep them performant, safe and useful for actual work without air-gapping them is by installing Linux on them, and OTD already supports Linux-on-x86-64 just fine. If you want to properly support macOS, please get with the programme and start fully and officially supporting Apple Silicon with an actual public release, STAT. You may be mad at Apple for always making these aggressive transitions and now even pestering your users with a nag alert on every reboot/user login, you may be mad at Apple users for shelling out big bucks for their machines and enabling them, you may want to shoot the messenger (yours truly), but that won't change these facts.

**Comment by Kuuuube** [comment]
Status update on this for anyone curious.

With https://github.com/OpenTabletDriver/OpenTabletDriver/pull/4943 being merged as well as https://github.com/OpenTabletDriver/OpenTabletDriver/pull/4721 and https://github.com/OpenTabletDriver/OpenTabletDriver/pull/4730 for signing support, we currently do have support for ARM/Apple Silicon macOS builds. It isn't piped into the CI or releases at the moment but it is available and working in the build system. There is no need to worry about not being able to use OTD after Apple decides to drop emulation support. We will have builds ready in time for the date they've set but even if they jump their deadline, we are set up to provide dev builds or allow users to easily build it themselves.

Apple has stated that intel app emulation will be [discontinued in macOS 28](https://support.apple.com/en-us/102527) and based on prior release history this version should be expected to release around September 2027.
